### PR TITLE
speech recognition sample

### PR DESCRIPTION
**Pull request by spazewalker**
#### Project details
* Mentor : @l-bat 
* Project proposal : https://summerofcode.withgoogle.com/projects/#5148521881141248

##### Creating ONNX model
NVIDIA trained jasper using FP16 precision. OpenCV needs FP32. We need to change onnx model's graph. This is done using this script : [convert_jasper_to_FP32.py](https://gist.github.com/spazewalker/507f1529e19aea7e8417f6e935851a01). Pre-trained converted onnx can be found [here](https://drive.google.com/drive/folders/1wLtxyao4ItAg8tt4Sb63zt6qXzhcQoR6?usp=sharing). Original pre-trained model by NVIDIA can be found [here](https://ngc.nvidia.com/catalog/models/nvidia:jasper_pyt_onnx_fp16_amp/files).

#### Usage
```bash
usage: speech_recognition.py [-h] --input_audio INPUT_AUDIO [--show_spectrogram] [--model MODEL] [--output OUTPUT] [--backend {0,2,3}] [--target {0,1,2}]

This script runs Jasper Speech recognition model

optional arguments:
  -h, --help            show this help message and exit
  --input_audio INPUT_AUDIO
                        Path to input audio file. OR Path to a txt file with relative path to multiple audio files in different lines (default: None)
  --show_spectrogram    Whether to show a spectrogram of the input audio. (default: False)
  --model MODEL         Path to the onnx file of Jasper. default="jasper.onnx" (default: jasper.onnx)
  --output OUTPUT       Path to file where recognized audio transcript must be saved. Leave this to print on console. (default: None)
  --backend {0,2,3}     Select a computation backend: 0: automatically (by default) 2: OpenVINO Inference Engine 3: OpenCV Implementation (default: 0)
  --target {0,1,2}      Select a target device: 0: CPU target (by default) 1: OpenCL 2: OpenCL FP16 (default: 0)
```

#### Todo
- [ ] Use AudioIO instead of soundfile.
- [ ] Check performance.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Docs
```

### CODE REVIEW

File: samples/dnn/speech_recognition.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    net.setPreferableTarget(args.target)
+    
+    # Get Filterbank Features
+    feature_extractor=FilterbankFeatures()
+    features = feature_extractor.calculate_features(x=X,seq_len=seq_len)
+
+    # Show spectogram if required
+    if args.show_spectrogram:
+        img = plt.imshow((features[0]), origin='lower', cmap='inferno', interpolation='nearest', aspect='auto')
+        plt.show()
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:390]: Can you use OpenCV instead of  `plt`?

File: samples/dnn/speech_recognition.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
+import numpy as np
+import cv2 as cv
+import argparse, os
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:3]:
```suggestion
import argparse
import os
```

**Comment by l-bat** [comment]
Please add description at the beginning of sample as in https://github.com/opencv/opencv/blob/b74aae6ca353c667f1dbfba615d7ef18bcdbb349/samples/dnn/human_parsing.py#L2-L40

1. How to get FP32 ONNX model from pre-trained model
2. Provide link to the converted model

File: samples/dnn/speech_recognition.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                decoded_prediction.append(p)
+            previous = p
+        hypothesis = ''.join([self.labels_map[c] for c in decoded_prediction])
+        hypotheses.append(hypothesis)
+        return hypotheses
+
+if __name__ == '__main__':
+
+    # Computation backends supported by layers
+    backends = (cv.dnn.DNN_BACKEND_DEFAULT, cv.dnn.DNN_BACKEND_OPENCV)
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:387]: Could you try forward net with OpenVINO (cv.dnn.DNN_BACKEND_INFERENCE_ENGINE)?

File: samples/dnn/speech_recognition.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+    # Target Devices for computation
+    targets = (cv.dnn.DNN_TARGET_CPU, cv.dnn.DNN_TARGET_OPENCL)
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:389]: Could you please add `cv.dnn.DNN_TARGET_OPENCL_FP16`?

File: samples/dnn/speech_recognition.py (hunk 3 above)
Comment (spazewalker) [reply samples/dnn/speech_recognition.py:387]: I tried. It gave this error: `error: (-213:The function/feature is not implemented) Unknown backend identifier in function 'cv::dnn::dnn4_v20210301::wrapMat'`

File: samples/dnn/speech_recognition.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+if __name__ == '__main__':
+
+    # Computation backends supported by layers
+    backends = (cv.dnn.DNN_BACKEND_DEFAULT, cv.dnn.DNN_BACKEND_HALIDE, cv.dnn.DNN_BACKEND_INFERENCE_ENGINE, cv.dnn.DNN_BACKEND_OPENCV)
+    # Target Devices for computation
+    targets = (cv.dnn.DNN_TARGET_CPU, cv.dnn.DNN_TARGET_OPENCL, cv.dnn.DNN_TARGET_OPENCL_FP16)
+
+    parser = argparse.ArgumentParser(description='This script runs Jasper Speech recognition model',
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--input_audio', type=str, help='Path to input audio file.')
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:391]: Do we need to specify supported audio formats?

File: samples/dnn/speech_recognition.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+    parser.add_argument('--show_spectrogram', action='store_true', help='Whether to show a spectrogram of the input audio.')
+    parser.add_argument('--model', type=str, default='jasper.onnx', help='Path to the onnx file of Jasper. default="jasper.onnx"')
+    parser.add_argument('--output', type=str, help='Path to file where recognized audio transcript must be saved. Leave this to print on console.')
+    parser.add_argument('--backend', choices=backends, default=cv.dnn.DNN_BACKEND_DEFAULT, type=int,
+                        help='Select a computation backend: '
+                        "%d: automatically (by default) "
+                        "%d: Halide"
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:398]: I'm not sure that Halide backend supports Jasper

File: samples/dnn/speech_recognition.py (hunk 5 above)
Comment (l-bat) [reply samples/dnn/speech_recognition.py:391]: I think we need to add `required=True`

File: samples/dnn/speech_recognition.py [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+     6. Finally save the model
+        ```
+        with open('jasper_dynamic_input_float.onnx','wb') as f:
+            onnx.save_model(model,f)
+        ```
+
+    Original Repo : https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/SpeechRecognition/Jasper
+ '''
+
+class FilterbankFeatures():
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:57]: No parentheses needed

File: samples/dnn/speech_recognition.py (hunk 5 above)
Comment (spazewalker) [reply samples/dnn/speech_recognition.py:391]: Finally, we need to use AudioIO. So, should I add the formats supported there? I suppose mp3, wav and mp4 are supported.

File: samples/dnn/speech_recognition.py (hunk 6 above)
Comment (spazewalker) [reply samples/dnn/speech_recognition.py:398]: I forgot to remove this. I was checking if it does.

File: samples/dnn/speech_recognition.py [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+
+    # Initialize decoder
+    decoder = Decoder()
+
+    # make prediction
+    net.setInput(features)
+    output = net.forward()
+
+    # decode output to transcript
+    prediction = decoder.decode(output[0])
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:445]: Why output[0] is needed?
Is the first dimension always equal to 1? For any batch size?

Comment (spazewalker) [reply samples/dnn/speech_recognition.py:445]: Since this script handles a single audio file, which is treated as batch size 1, the input to jasper is always of shape (1,64,?) and gives an output of shape (1,?,29).

Comment (l-bat) [reply samples/dnn/speech_recognition.py:445]: Could you please check how model works with several audio files (batch size >1)?

File: samples/dnn/speech_recognition.py [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+    if args.input_audio and not os.path.isfile(args.input_audio):
+        raise OSError("Input audio file does not exist")
+    if not os.path.isfile(args.model):
+        raise OSError("Jasper model file does not exist")
+
+    # Read audio File
+    try:
+        audio = sf.read(args.input_audio)
+        X=audio[0]
+        seq_len=np.array([X.shape[0]],dtype=np.int32)
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:417]:
```suggestion
        assert audio.shape[0] == 1
        X = audio[0]
        seq_len = np.array([X.shape[0]], dtype=np.int32)
```

File: samples/dnn/speech_recognition.py [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+        raise Exception(f"Soundfile cannot read {args.input_audio}. Try a different format")
+
+    # Load Network
+    net = cv.dnn.readNetFromONNX(args.model)
+    net.setPreferableBackend(args.backend)
+    net.setPreferableTarget(args.target)
+
+    # Get Filterbank Features
+    feature_extractor = FilterbankFeatures()
+    features = feature_extractor.calculate_features(x=X,seq_len=seq_len)
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:428]:
```suggestion
    features = feature_extractor.calculate_features(x=X, seq_len=seq_len)
```

File: samples/dnn/speech_recognition.py [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+
+    # Show spectogram if required
+    if args.show_spectrogram:
+        img = cv.normalize(src=features[0], dst=None, alpha=0, beta=255, norm_type=cv.NORM_MINMAX, dtype=cv.CV_8U)
+        img = cv.applyColorMap(img,cv.COLORMAP_JET)
+        cv.imshow('spectogram',img)
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:434]:
```suggestion
        img = cv.applyColorMap(img, cv.COLORMAP_JET)
        cv.imshow('spectogram', img)
```

File: samples/dnn/speech_recognition.py [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+    output = net.forward()
+
+    # decode output to transcript
+    prediction = decoder.decode(output[0])
+
+    # save transcript if required
+    if args.output:
+        with open(args.output,'w') as f:
+            f.write(prediction)
+        print("Done")
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:451]:
```suggestion
        print("Transcript was written to {}".format(args.output))
```

File: samples/dnn/speech_recognition.py [hunk 13]
```diff
@@ -0,0 +1,506 @@
...
+    # expanding 1 dimentions allows us to pass it to the network
+    # from python. This should be resolved in the future.
+    features = np.expand_dims(features,axis=3)
+
+    # make prediction
+    net.setInput(features)
+    output = net.forward()
+
+    # decode output to transcript
+    prediction = decoder.decode(output[0])
```
Comment (l-bat) [inline samples/dnn/speech_recognition.py:412]: What do you think about replace it to 
```suggestion
    prediction = decoder.decode(output.squeeze(0))
```
?

Comment (spazewalker) [reply samples/dnn/speech_recognition.py:412]: that's better

Comment (spazewalker) [reply samples/dnn/speech_recognition.py:412]: Let me commit it here. There are a lot of commits in this PR at this point. I'll rebase it before finalizing.

**Comment by alalek** [comment]
@spazewalker Could you please check if approach from #20558 works for this case?

**Comment by spazewalker** [comment]
> @spazewalker Could you please check if approach from #20558 works for this case?

@alalek Just tested it. It works for this case.

**Comment by alalek** [comment]
Lets merge it with `soundfile` workaround.

@spazewalker Please make PR to "Ready for review" if it is ready for merging.

**Comment by alalek** [comment]
> "Ready for review"

@spazewalker Ping. Or let us know if you want to improve something else.

**Comment by spazewalker** [comment]
@alalek I'm actually waiting for #19721 to get merged. I think videoio would replace the soundfile.

**Review by alalek — APPROVED** [review]
Thank you :+1:
